### PR TITLE
DOC: draft of release notes

### PR DESCRIPTION
**Pull request by tacaswell**

### CODE REVIEW

File: doc/source/release_notes.rst [hunk 1]
```diff
@@ -2,6 +2,64 @@
  Release History
 =================
 
+v1.4.0 (2019-06-XX)
+===================
+
+Features
+--------
+* Version ohpyd classes to support IOCs changing over time. This is currently
```
Comment (klauer) [inline doc/source/release_notes.rst]:
```suggestion
* Version ophyd classes to support IOCs changing over time. This is currently
```

File: doc/source/release_notes.rst [hunk 2]
```diff
@@ -2,6 +2,64 @@
...
+  implemented from AreaDetector and it's plugins (support for 1-9 to 3-5).
```
Comment (klauer) [inline doc/source/release_notes.rst]:
```suggestion
  implemented for AreaDetector plugins (releases 1-9 to 3-4).
```

File: doc/source/release_notes.rst [hunk 3]
```diff
@@ -2,6 +2,64 @@
...
+* `.NDDerivedSignal` to expose a re-shaped version of an N-D array.
```
Comment (klauer) [inline doc/source/release_notes.rst]:
```suggestion
* Added `NDDerivedSignal`, which supplies a reshaped version of an input array signal.
```

File: doc/source/release_notes.rst [hunk 4]
```diff
@@ -2,6 +2,64 @@
...
+* `~.ophyd.device.Component.subscriptions` decorator to set up default
```
Comment (klauer) [inline doc/source/release_notes.rst]:
```suggestion
* Added `~.ophyd.device.Component.subscriptions` decorator to set up default
```

File: doc/source/release_notes.rst [hunk 5]
```diff
@@ -2,6 +2,64 @@
...
+  subscriptions on a component.
+* Added `.create_device_from_components` to aid dynamic device generation.
+* Added `.required_for_connection` decorator to mark the Components that must
+  be connected for the overall Device to be considered connected.
+* Added a hook to be called on all ophyd object creation via the
+  `.ohpydobj.add_instantiation_callback` class method along with the
```
Comment (klauer) [inline doc/source/release_notes.rst]:
```suggestion
  `.ophydobj.add_instantiation_callback` class method along with the
```

File: doc/source/release_notes.rst [hunk 6]
```diff
@@ -2,6 +2,64 @@
...
+* Added `.ophydobj.dotted_name` property to get the full attribute
+  name of a child component.
+* Added ``read_access`` and ``write_access`` properties.
+* Added ``destroy`` method to all ohpyd objects.
+
+API Changes
+-----------
+
+* Drop support for Python3.5
+* If using pyepics, requires pyepics >= 3.4.0
```
Comment (klauer) [inline doc/source/release_notes.rst]:
```suggestion
* The optional dependency pyepics, if installed, must be above version 3.4.0 to be used.
```

File: doc/source/release_notes.rst [hunk 7]
```diff
@@ -2,6 +2,64 @@
...
+* If you are holding a reference to a pyepics.pv.PV that is shared
+  with ophyd and all ophyd objects that use that PV are torn down, all
+  callbacks on the PV will be cleared and the channel will be torn
+  down.  If this is a problem for you, please create a bug report.
+* Removed ``ophyd.control_layer``.  The "control layer" used to access
+  EPICS can be controlled via `~ophyd.set_cl`.
+* `.DynamicDeviceComponent` is now a `.Component` sub-class
+* Change argument name in `.ophydobj.subscribe` from ``cb`` to
```
Comment (klauer) [inline doc/source/release_notes.rst]:
```suggestion
* Changed argument name in `.ophydobj.subscribe` from ``cb`` to
```

File: doc/source/release_notes.rst [hunk 8]
```diff
@@ -2,6 +2,64 @@
...
+  ``callback``.
+* Removed ``ophyd.tests.conftest.FakeEpicsPV``,
+  ``ophyd.tests.conftest.FakeEpicsWaveForm`` , and associated helper
+  functions.  If you need this class, please vendor it from a previous
+  version of ophyd.
```
Comment (klauer) [inline doc/source/release_notes.rst]:
```suggestion
  version of ophyd - or please consider moving to `make_fake_device` or `caproto`-based simulation IOCs.
```

File: doc/source/release_notes.rst [hunk 9]
```diff
@@ -2,6 +2,64 @@
...
+* Removed ``ophyd.tests.AssertTolls`` and use standard pytest idioms
```
Comment (klauer) [inline doc/source/release_notes.rst]:
```suggestion
* Removed ``ophyd.tests.AssertTools`` and use standard pytest idioms
```

File: doc/source/release_notes.rst [hunk 10]
```diff
@@ -2,6 +2,64 @@
...
+  throughout test suite
+
+
+Internals
+---------
+* switch from `.Device` using meta-classes to using a
+  ``__init__subclass__``.  There are now no meta-classes in ohpyd!
```
Comment (klauer) [inline doc/source/release_notes.rst]:
```suggestion
  ``__init__subclass__``.  There are now no meta-classes in ophyd!
```

File: doc/source/release_notes.rst [hunk 11]
```diff
@@ -2,6 +2,64 @@
...
+* Completely overhauled how we connect to PVs at initialization and tear
+  them down on destruction of the ophyd objects.
+* Completely overhauled how PV meta-data is handled.
+* Completely overhauled the dispatcher thread mechanism.
+* Removed our backport of enums.
```
Comment (klauer) [inline doc/source/release_notes.rst]:
```suggestion
* Removed our backport of `enum`, as it is available on the minimum Python 3.6.
```

File: doc/source/release_notes.rst [hunk 12]
```diff
@@ -2,6 +2,64 @@
...
+* Added `.required_for_connection` decorator to mark the Components that must
+  be connected for the overall Device to be considered connected.
+* Added a hook to be called on all ophyd object creation via the
+  `.ohpydobj.add_instantiation_callback` class method along with the
+  `.register_instances_keyed_on_name` and
+  `.register_instance_in_weakset` helper functions.
+* Added `.ophydobj.dotted_name` property to get the full attribute
+  name of a child component.
+* Added ``read_access`` and ``write_access`` properties.
+* Added ``destroy`` method to all ohpyd objects.
```
Comment (klauer) [inline doc/source/release_notes.rst]:
```suggestion
* Added ``destroy`` method to all ophyd objects.
```

**Comment by tacaswell** [comment]
Thank you for the through review @klauer !

**Comment by danielballan** [comment]
The nightly build has failed. Shouldn't both nightly builds be in the allowed failures block? One is and one is not.

**Comment by klauer** [comment]
That nightly failed in quite an extraordinary way. Yes, I think all nightlies should be allowed failures.

File: doc/source/release_notes.rst [hunk 13]
```diff
@@ -2,6 +2,64 @@
...
+  with ophyd and all ophyd objects that use that PV are torn down, all
+  callbacks on the PV will be cleared and the channel will be torn
+  down.  If this is a problem for you, please create a bug report.
+* Removed ``ophyd.control_layer``.  The "control layer" used to access
+  EPICS can be controlled via `~ophyd.set_cl`.
+* `.DynamicDeviceComponent` is now a `.Component` sub-class
+* Changed argument name in `.ophydobj.subscribe` from ``cb`` to
+  ``callback``.
+* Removed ``ophyd.tests.conftest.FakeEpicsPV``,
+  ``ophyd.tests.conftest.FakeEpicsWaveForm`` , and associated helper
```
Comment (mrakitin) [inline doc/source/release_notes.rst]:
```suggestion
  ``ophyd.tests.conftest.FakeEpicsWaveForm``, and associated helper
```

File: doc/source/release_notes.rst [hunk 14]
```diff
@@ -2,6 +2,64 @@
...
+  functions.  If you need this class, please vendor it from a previous
+  version of ophyd - or please consider moving to `make_fake_device` or `caproto`-based simulation IOCs.
+* Removed ``ophyd.tests.AssertTools`` and use standard pytest idioms
+  throughout test suite
```
Comment (mrakitin) [inline doc/source/release_notes.rst]:
```suggestion
  throughout test suite.
```

File: doc/source/release_notes.rst [hunk 15]
```diff
@@ -2,6 +2,64 @@
...
+
+
+Internals
+---------
+* switch from `.Device` using meta-classes to using a
```
Comment (mrakitin) [inline doc/source/release_notes.rst]:
```suggestion
* Switch from `.Device` using meta-classes to using a
```

File: doc/source/release_notes.rst [hunk 16]
```diff
@@ -2,6 +2,64 @@
...
+  down.  If this is a problem for you, please create a bug report.
+* Removed ``ophyd.control_layer``.  The "control layer" used to access
+  EPICS can be controlled via `~ophyd.set_cl`.
+* `.DynamicDeviceComponent` is now a `.Component` sub-class
+* Changed argument name in `.ophydobj.subscribe` from ``cb`` to
+  ``callback``.
+* Removed ``ophyd.tests.conftest.FakeEpicsPV``,
+  ``ophyd.tests.conftest.FakeEpicsWaveForm`` , and associated helper
+  functions.  If you need this class, please vendor it from a previous
+  version of ophyd - or please consider moving to `make_fake_device` or `caproto`-based simulation IOCs.
```
Comment (mrakitin) [inline doc/source/release_notes.rst]:
```suggestion
  version of ophyd - or please consider moving to `make_fake_device`
  or `caproto`-based simulation IOCs.
```

File: doc/source/release_notes.rst [hunk 17]
```diff
@@ -2,6 +2,64 @@
...
+* Completely overhauled how we connect to PVs at initialization and tear
+  them down on destruction of the ophyd objects.
+* Completely overhauled how PV meta-data is handled.
+* Completely overhauled the dispatcher thread mechanism.
+* Removed our backport of `enum`, as it is available on the minimum Python 3.6.
+
+
+
+
+
```
Comment (mrakitin) [inline doc/source/release_notes.rst]: Do we need that many blank spaces?

Comment (klauer) [reply doc/source/release_notes.rst]: Definitely not!

**Comment by klauer** [comment]
I couldn't accept your suggestions, @mrakitin, despite being able to push to @tacaswell's branch. I accepted them manually, and then did another pass at it. The sphinx tags were not being rendered correctly _at all_, so they needed to be redone...

**Comment by mrakitin** [comment]
Thanks, @klauer! I am happy with the changes. I will keep my review in the "blocking" mode so we don't forget to bump the release date.

**Pull request merged**